### PR TITLE
Verify ics23 proofs

### LINKED ISSUE #347: Integrate ICS23 proof validation for state queries in client
(not including header validation)

- Spec: https://github.com/cosmos/ics/tree/master/spec/ics-023-vector-commitments
- Implementation: https://github.com/confio/ics23

to be integrated in `packages/stargate/src/stargateclient.ts`:

```ts
  private async queryVerified(store: string, key: Uint8Array): Promise<Uint8Array> {
    const response = await this.tmClient.abciQuery({
      // we need the StoreKey for the module, not the module name
      // https://github.com/cosmos/cosmos-sdk/blob/8cab43c8120fec5200c3459cbf4a92017bb6f287/x/auth/types/keys.go#L12
      path: `/store/${store}/key`,
      data: key,
      prove: true,
    });

    if (response.code) {
      throw new Error(`Query failed with (${response.code}): ${response.log}`);
    }

    if (!arrayContentEquals(response.key, key)) {
      throw new Error(`Response key ${toHex(response.key)} doesn't match query key ${toHex(key)}`);
    }

    // TODO: implement proof verification
    // https://github.com/CosmWasm/cosmjs/issues/347

    return response.value;
  }
```

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #347 

- [x] Parses proof info in tendermint rpc
- [x] Validate proper sdk proof format in `queryVerified`
- [x] Validate the ics23 proof itself

### CODE REVIEW

**Comment by ethanfrey** [comment]
@webmaster128 I am almost there.

The missing step is "get the header and verify with appHash". The issue is for a query that returns state after block H, I need the appHash from block H+1. There are 2 issues here:

1. We need a cleaner way than `sleep(5000)` to wait for the next header.
2. I don't see any header queries on tendermintRpc, just the block query.

Maybe I am missing something. With some tips on the tendermint rpc header stuff, I am pretty sure I can get this all finished - no significant roadblocks yet

**Comment by ethanfrey** [comment]
The work is done. It can probably be refactored and use some more targeted tests.

Let me know what improvements you would suggest

File: packages/stargate/src/queries/queryclient.ts [hunk 1]
```diff
@@ -1,8 +1,45 @@
 /* eslint-disable no-dupe-class-members, @typescript-eslint/ban-types, @typescript-eslint/naming-convention */
-import { toHex } from "@cosmjs/encoding";
+import { ics23, verifyExistence, verifyNonExistence } from "@confio/ics23";
+import { fromAscii, toAscii, toHex } from "@cosmjs/encoding";
+import { firstEvent } from "@cosmjs/stream";
 import { Client as TendermintClient } from "@cosmjs/tendermint-rpc";
 import { arrayContentEquals, assert, isNonNullObject } from "@cosmjs/utils";
 
+// TODO: import these from @confio/ics23 when exported
```
Comment (webmaster128) [inline packages/stargate/src/queries/queryclient.ts:8]: Do you create an update for this?

File: packages/stargate/src/queries/queryclient.ts [hunk 2]
```diff
@@ -1,8 +1,45 @@
...
+export const IavlSpec: ics23.IProofSpec = {
```
Comment (webmaster128) [inline packages/stargate/src/queries/queryclient.ts:9]: This should be lower case

File: packages/stargate/src/queries/queryclient.ts [hunk 3]
```diff
@@ -1,8 +1,45 @@
...
+  innerSpec: {
+    childOrder: [0, 1],
+    minPrefixLength: 4,
+    maxPrefixLength: 12,
+    childSize: 33,
+    hash: ics23.HashOp.SHA256,
+  },
+};
+
+export const TendermintSpec: ics23.IProofSpec = {
```
Comment (webmaster128) [inline packages/stargate/src/queries/queryclient.ts:26]: here too

File: packages/tendermint-rpc/src/v0-33/responses.ts [hunk 4]
```diff
@@ -62,7 +82,7 @@ function decodeAbciQuery(data: RpcAbciQueryResponse): responses.AbciQueryRespons
   return {
     key: Base64.decode(optional(data.key, "" as Base64String)),
     value: Base64.decode(optional(data.value, "" as Base64String)),
-    // proof: may(Base64.decode, data.proof),
+    proof: may(decodeQueryProof, data.proof),
```
Comment (webmaster128) [inline packages/tendermint-rpc/src/v0-33/responses.ts:85]: I think we should test this field in a plain Tendermint test. Is this possible with our testing `--proxy_app=kvstore`?

**Comment by webmaster128** [comment]
> I don't see any header queries on tendermintRpc, just the block query.

`tmClient.blockchain` is probably what you are looking for

File: packages/tendermint-rpc/src/v0-33/responses.ts (hunk 4 above)
Comment (ethanfrey) [reply packages/tendermint-rpc/src/v0-33/responses.ts:85]: There are no proofs there. We can get undefined.

The actual format we parse is only from the cosmos-sdk

**Comment by ethanfrey** [comment]
> tmClient.blockchain is probably what you are looking for

Ah, good stuff.  I will look there. The `nextHeader` subscription worked out quite well actually.

**Comment by webmaster128** [comment]
> The `nextHeader` subscription worked out quite well actually.

This makes the whole Query client depend on a WebSocket connection. It would be better to avoid requiring that. It might be nice as a speed optimization.

File: packages/stargate/src/queries/queryclient.ts (hunk 1 above)
Comment (ethanfrey) [reply packages/stargate/src/queries/queryclient.ts:8]: Published https://www.npmjs.com/package/@confio/ics23/v/0.6.3

**Comment by ethanfrey** [comment]
I made all requested changes and some other cleanup.

> This makes the whole Query client depend on a WebSocket connection. It would be better to avoid requiring that. It might be nice as a speed optimization.

I tried with the other query and it doesn't work, I need to wait on the chain. Do you really want a polling loop there? (See https://github.com/CosmWasm/cosmjs/pull/386/commits/51d07b3f154d20097596b2e7a2e735db352e1d09)

**Comment by webmaster128** [comment]
The CI fails because of https://github.com/CosmWasm/cosmjs/pull/394



> I tried with the other query and it doesn't work, I need to wait on the chain. Do you really want a polling loop there? (See [51d07b3](https://github.com/CosmWasm/cosmjs/commit/51d07b3f154d20097596b2e7a2e735db352e1d09))

What I really want i a query that can be verified without waiting seconds for the next block. But if we don't want to give up HTTP support, we need a polling fallback. I implemented one in #395. Once this is merged in here, this is good to go from my point of view.

**Comment by ethanfrey** [comment]
> What I really want i a query that can be verified without waiting seconds for the next block. 

This is not possible with the sdk / tendermint design. Block H includes a list of transactions *to be executed* and the AppHash resulting from *executing the transactions in H-1*.

The only way to do this without a sleep/wait is to query the previous block. So, if we get header H as the most recent, then query on height H-1. This may be a bit surprising to a user, as it may give a different result than an unverified query, but also is much faster (no waiting for a block). We just need to document the approach and then do it.

**Comment by ethanfrey** [comment]
Rebased on #394 to fix test error

**Comment by webmaster128** [comment]
> This is not possible with the sdk / tendermint design.

This is what I suspected. Let's just keep subscribe or poll depending on what is available

**Comment by webmaster128** [comment]
I'll finalize and merge this now

**Pull request merged**